torch/nested/_internal/nested_tensor.py

```diff
@@ -0,0 +1,161 @@
+import torch
+from typing import Tuple
+from torch._C import DispatchKey, DispatchKeySet
+from torch.utils.weak import WeakTensorKeyDictionary
+from typing import *  # noqa: F403
+
+_tensor_id_counter = 0
+_tensor_id_registry = WeakTensorKeyDictionary()
+
+def get_tensor_id(tensor):
+    global _tensor_id_counter
+    if tensor not in _tensor_id_registry:
+        _tensor_id_registry[tensor] = _tensor_id_counter
+        _tensor_id_counter += 1
+    return torch._C._get_singleton_int(_tensor_id_registry[tensor])
+
+class NestedTensor(torch.Tensor):
+    buffer: torch.Tensor
+    offsets: torch.Tensor
```

Comment:
as discussed, we should name this smth like `offsets_` and introduce an `offsets()` accessor on the subclass for consistency with `values()` / `indices()`